edb/pgsql/delta.py

```diff
@@ -2252,36 +2280,337 @@ def adjust_pointer_storage(self, pointer, schema, orig_schema, context):
+
+        if alters:
+            assert len(alters) == 1, 'expected only one SET TYPE'
+            using_eql_expr = next(iter(alters)).expr
+        else:
+            using_eql_expr = None
+
+        # For links, when the new type is a supertype of the old, no
+        # SQL-level changes are necessary, unless an explicit conversion
+        # expression was specified.
+        if (
+            is_link
+            and using_expr is None
+            and orig_target.issubclass(schema, new_target)
+        ):
+            return
+
+        if using_eql_expr is not None:
+            # The EdgeQL statement contained a USING clause.  There are
+            # two major possibilities: 1) trivial case, where the USING
+            # clause refers only to the columns of the source table, in
+            # which case we simply compile that into an equivalent SQL
+            # USING clause, and 2) complex case, which supports arbitrary
+            # queries, but requires a temporary column, which is populated
+            # with the transition query and then used as the source for
+            # the SQL USING clause.
+            if using_eql_expr.irast is not None:
+                ir = using_eql_expr.irast
+            else:
+                using_eql_expr = self._compile_expr(
+                    orig_schema,
+                    context,
+                    using_eql_expr,
+                    target_as_singleton=True,
+                )
+                ir = using_eql_expr.irast
+
+            assert ir is not None
+
+            expr_is_nullable = using_eql_expr.cardinality.can_be_zero()
+
+            refs = irutils.get_longest_paths(ir.expr)
+            ref_tables = schemamech.get_ref_storage_info(ir.schema, refs)
+
+            local_table_only = all(
+                t == old_ptr_stor_info.table_name
+                for t in ref_tables
+            )
+
+            using_expr_is_trivial = (
+                not is_link
+                # SQL SET TYPE cannot contain references
+                # outside of the local table.
+                and local_table_only
+                # Changes to a multi-pointer might involve contraction of
+                # the overall cardinality, i.e. the deletion some rows.
+                and not is_multi
+                # If the property is required, and the USING expression
+                # was not proven by the compiler to not return ZERO, we
+                # must inject an explicit NULL guard, as the SQL null
+                # violation error is very nondescript in the context of
+                # a table rewrite, making it hard to pinpoint the failing
+                # object.
+                and (not is_required or not expr_is_nullable)
+            )
+
+            need_temp_col = (
+                (is_multi and expr_is_nullable)
+                or (changing_col_type and not using_expr_is_trivial)
+            )
+
+            if not using_expr_is_trivial:
```

Comment:
Add a comment about what's being set up here?

Reply:
The relevant comment is under the `if using_eql_expr is not None`, I'll add a pointer to it